<launch>

    <!-- SYSTEM MODEL -->
    <!-- vehicle parameters -->
    <param name="L_a" type="double" value="0.125" />
    <param name="L_b" type="double" value="0.125" />
    <param name="m" type="double" value="1.98" />
    <param name="I_z" type="double" value="0.24" />
    <param name="dt" type="double" value="0.1" />
    <param name="num_agents" type="int" value="1" />

    <!-- MACHINES -->
    <machine name="yoga" address="127.0.0.1" user="lukas" timeout="20.0" env-loader="/home/lukas/barc/env_loader_yoga.sh" />
    <!-- <machine name="yoga" address="192.168.10.74" user="lukas" timeout="20.0" env-loader="/home/lukas/barc/env_loader_yoga.sh" /> -->
    <machine name="this-pc" address="192.168.10.147" timeout="20.0" env-loader="/home/mpcubuntu/barc/env_loader_pc.sh"/>

    <machine name="odroid" address="192.168.10.213" user="odroid" timeout="20.0" env-loader="/home/odroid/barc/env_loader_odroid.sh"/>

    <!-- AGENT 1 NODES-->
    <!-- AGENT-NODE 1-->
    <node machine="yoga" pkg="barc" type="agent_node.jl" name="agent_1" output="screen">
        <param name="index" type="int" value="1" />
        <param name="color" type="str" value="blue" />
        <param name="mass" type="double" value="1.98" />
        <param name="time_offset" type="double" value="0.0" />
        <param name="delay_a" type="int" value="1" />
        <param name="delay_df" type="int" value="1" />

        <remap from="pos_info" to="agent_1/pos_info" />
        <remap from="ecu" to="agent_1/ecu" />
        <remap from="xy_prediction" to="agent_1/xy_prediction" />
        <remap from="selected_states" to="agent_1/selected_states" />
        <remap from="prediction" to="agent_1/prediction" />
        <remap from="adv_prediction" to="agent_2/prediction" />
        <remap from="ecu_pwm" to="agent_1/ecu_pwm" />
    </node>
    
    <!--DYNAMIC OBSERVER-->
    <node machine="yoga" pkg="barc" type="state_estimation_SensorKinematicModel.py" name="state_estimation_dynamic_1" output="screen">
        <!-- ekf properties -->
        <param name="q_std" type="double" value="0.1" />
        <param name="psi_std" type="double" value="0.5" />
        <param name="v_std" type="double" value="0.5" />
        <param name="gps_std" type="double" value="0.8" />
        <param name="ang_v_std" type="double" value="0.8" />
        <param name="est_mode" type="double" value="1" />

        <remap from="hedge_pos" to="agent_1/hedge_pos" />
        <remap from="imu/data" to="agent_1/imu/data" />
        <remap from="vel_est" to="agent_1/vel_est" />
        <remap from="pos_info" to="agent_1/pos_info" />
        <remap from="ecu" to="agent_1/ecu" />
        <remap from="hedge_imu_fusion" to="agent_1/hedge_imu_fusion" />
    </node>

    <!-- IMU -->
    <node machine="odroid" pkg="myahrs_driver" type="myahrs_driver" name="myahrs_driver">
        <param name="port" value="/dev/ttyACM0" />
        <param name="baud_rate" value="115200" />

        <remap from="imu/data" to="agent_1/imu/data" />
        <remap from="imu/data_raw" to="agent_1/imu/data_raw" />
        <remap from="imu/mag" to="agent_1/imu/mag" />
        <remap from="imu/temperature" to="agent_1/imu/temperature" />
        <remap from="tf" to="agent_1/tf" />
    </node>

    <!-- GPS -->
    <node machine="odroid" pkg="marvelmind_nav" type="hedge_rcv_bin" name="indoor_gps" args="/dev/ttyACM1" >
        <remap from="beacons_pos_a" to="agent_1/beacons_pos_a" />
        <remap from="hedge_pos" to="agent_1/hedge_pos" />
        <remap from="hedge_pos_a" to="agent_1/hedge_pos_a" />
        <remap from="hedge_imu_fusion" to="agent_1/hedge_imu_fusion" />
    </node>
    
    <!-- ARDUINO -->
    <!--  * encoders and ultrasound sensors -->
    <node machine="odroid" pkg="rosserial_python" type="serial_node.py" name="arduino_node" >
        <param name="port" value="/dev/ttyUSB0" />
        <remap from="vel_est" to="agent_1/vel_est" />
        <remap from="ecu_pwm" to="agent_1/ecu_pwm" />
        <remap from="diagnostics" to="agent_1/diagnostics" />
    </node>
    <!-- AGENT 1 NODES-->

    <!-- VISUALIZATION -->
    <node machine="yoga" pkg="barc" type="plotter.py" name="visualization" output="screen">
        <param name="num_agents" type="int" value="1" />
        <param name="index" type="int" value="1" />
    </node>

</launch>
